rules/aip0121/no_mutable_cycles.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0121
+
+import (
+	"bitbucket.org/creachadair/stringset"
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+	rpb "google.golang.org/genproto/googleapis/api/annotations"
+)
+
+var noMutableCycles = &lint.MessageRule{
+	Name:   lint.NewRuleName(121, "no-mutable-cycles"),
+	OnlyIf: utils.IsResource,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		var problems []lint.Problem
+		res := utils.GetResource(m)
+
+		for _, field := range m.GetFields() {
```

Comment:
nit: you could refactor this whole loop into findCycle, I think.

It looks like this logic exists just to help get the original resource reference. But there's nothing special about the original resource. You could just return whatever field is involved in the cycle, and as long as your problem enumerates all the seen resources, you could fix one of those chains in the cycle.

so basically if you get rid of orig and just use seen, you can get similar behavior.

Reply:
Yeah, let me stew on this a bit, trying to remember why I didn't push this into `findCycle` in the first place.

Reply:
Ok fixed it! Done!

It was just a tad tricky because we wanted to ensure that the original message's field that started the chain ultimately ended up as the descriptor/location on the finding, but we want to lint all of the fields in the original resource and only the one that creates a cycle in the recursive call. Figured it out though and added a test that involves multiple cycles on one resource across multiple fields, one deep and one shallow.